### PR TITLE
FOGL-7658 Control pipeline documentation & core support

### PR DESCRIPTION
**Pull request by MarkRiddoch**

### CODE REVIEW

**Review by pintomax — APPROVED** [review]

File: docs/control.rst [hunk 1]
```diff
@@ -380,3 +398,159 @@ Advanced Configuration
...
+The user has a number of actions that may be taken from this screen.
+
+  - Enable or disable the pipeline by clicking on the checkbox to the left of the pipeline name
+
+  - Click on the name of the plugin to view and edit the pipeline.
+
+  - Click on the three vertical dots to view the content menu.
+
+    +-------------------------+
+    | |pipeline_context_menu| |
```
Comment (ashish-jabble) [inline docs/control.rst:502]: Undefined substitution referenced.
It should define above in images link section.

File: docs/control.rst [hunk 2]
```diff
@@ -380,3 +398,159 @@ Advanced Configuration
...
+The Fledge Graphical User Interface provides a mechanism to manage the control pipelines, this is found in the Control sub menu under the pipelines item.
+
++-----------------+
+| |pipeline_menu| |
++-----------------+
+
+The user is presented with a list of the pipelines that have been created to date and an option in the top right corner to add a new pipeline.
+
++-----------------+
+| |pipeline_list| |
```
Comment (ashish-jabble) [inline docs/control.rst:488]: Don't you think we should have execution model column at GUI side to display its value? At the moment only with round corner badge

File: docs/control.rst [hunk 3]
```diff
@@ -380,3 +398,159 @@ Advanced Configuration
...
+  - **Source**: The control source that this pipeline should be considered to be used by.
+
+    +-------------------+
+    | |pipeline_source| |
+    +-------------------+
+
+  - **Filters**: The filters in the pipeline. Click on *Add new filter* to add a new filter to the pipeline.
+
+
+    Clicking on the *Add filter* link will display a dialogue in which the filter plugin can be chosen and named.
```
Comment (ashish-jabble) [inline docs/control.rst:531]: dialog

**Review by ashish-jabble — CHANGES_REQUESTED** [review]
There were some missing image links and other miscellaneous things.

File: docs/control.rst [hunk 4]
```diff
@@ -380,3 +398,159 @@ Advanced Configuration
...
+Pipeline Connections
+--------------------
+
+The control pipelines are not defined against a particular end point as they are with the data pipelines, they are defined separately and part of that definition includes the input and output end points to which the control pipeline may be attached. The input and output of a control pipeline may be defined as being able to connect to one of a set of endpoints.
+
++--------------+-------------+-----------------------------------------------------------------------------------------+
+| Type         | Endpoints   | Description                                                                             |
++==============+=============+=========================================================================================+
+| Any          | Both        | The pipeline can connection to any source or destination. This is only used in          |
+|              |             | situations where an exact match for an endpoint can not be satisfied.                   |
```
[the remark this answers is not in the document]
Comment (MarkRiddoch) [reply docs/control.rst:440]: Yes, it was meant to apply to both. Let's raise a separate Jira to change this but leave the documentation.

File: docs/control.rst [hunk 5]
```diff
@@ -380,3 +398,159 @@ Advanced Configuration
...
++--------------+-------------+-----------------------------------------------------------------------------------------+
+| API          | Source      | The source of the request is an API call the the public API of the Fledge instance.     |
```
[the remark this answers is not in the document]
Comment (MarkRiddoch) [reply docs/control.rst:442]: This is functionality that is not available yet, but the source would lie API and would include the name of the user.

File: docs/control.rst [hunk 6]
```diff
@@ -380,3 +398,159 @@ Advanced Configuration
...
++--------------+-------------+-----------------------------------------------------------------------------------------+
+
+Control pipelines are always executed in the control dispatcher service. When a request comes into the service it will look for a pipeline to pass that request through. This process will look at the source of the request and the destination of the request. If a pipeline that has source and destination endpoints that are an exact match for the source and destination of the control request then the control request will be processed through that pipeline.
+
+If no exact match is found then the source of the request will be checked against the defined pipelines for a match with the specified source and a destination of *any*. If there is a pipeline that matches these criteria it will be used. If not then a check is made for a pipeline with a source of *any* and a destination that matches the destination of this request.
+
+If all the above tests fail then a final test is made for a pipeline with a source of *any* and a destination of *any*. If no match occurs then the request is processed without passing through any filters.
```
[the remark this answers is not in the document]
Comment (MarkRiddoch) [reply docs/control.rst:461]: We need to update the API.

File: docs/control.rst (hunk 2 above)
Comment (MarkRiddoch) [reply docs/control.rst:488]: I think it is fine as it is.

File: docs/control.rst (hunk 4 above)
[the remark this answers is not in the document]
Comment (ashish-jabble) [reply docs/control.rst:440]: FOGL-7984 is created

File: docs/control.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (ashish-jabble) [reply docs/control.rst:461]: FOGL-7984 is created

File: docs/control.rst (hunk 1 above)
Comment (MarkRiddoch) [reply docs/control.rst:505]: Added definition of image

**Review by ashish-jabble — APPROVED** [review]

**Pull request merged**